src/VisualStudio/Core/Next/FindReferences/StreamingFindReferencesPresenter.TableDataSourceFindReferencesContext.cs

```diff
@@ -440,25 +447,73 @@ private TextSpan GetRegionSpanForReference(SourceText sourceText, TextSpan refer
                 // Fill in those gaps so we have *all* parts of the span 
                 // classified properly.
-                var syntaxParts = FillInClassifiedSpanGaps(sourceText, widenedSpan.Start, syntaxSpans);
-                var semanticParts = FillInClassifiedSpanGaps(sourceText, widenedSpan.Start, semanticSpans);
+                var filledInSyntaxSpans = ArrayBuilder<ClassifiedSpan>.GetInstance();
+                var filledInSemanticSpans = ArrayBuilder<ClassifiedSpan>.GetInstance();
+
+                try
+                {
+                    FillInClassifiedSpanGaps(sourceText, widenedSpan.Start, syntaxSpans, filledInSyntaxSpans);
+                    FillInClassifiedSpanGaps(sourceText, widenedSpan.Start, semanticSpans, filledInSemanticSpans);
 
-                // Now merge the lists together, taking all the results from syntaxParts
-                // unless they were overridden by results in semanticParts.
-                return MergeParts(syntaxParts, semanticParts);
+                    // Now merge the lists together, taking all the results from syntaxParts
+                    // unless they were overridden by results in semanticParts.
+                    return MergeParts(filledInSyntaxSpans, filledInSemanticSpans);
+                }
+                finally
+                {
+                    filledInSyntaxSpans.Free();
+                    filledInSemanticSpans.Free();
+                }
             }
 
-            private static List<ClassifiedSpan> FillInClassifiedSpanGaps(
-                SourceText sourceText, int startPosition, IEnumerable<ClassifiedSpan> classifiedSpans)
+            private void AdjustSpans(List<ClassifiedSpan> spans, TextSpan widenedSpan)
             {
-                var result = new List<ClassifiedSpan>();
+                for (var i = 0; i < spans.Count; i++)
+                {
+                    var span = spans[i];
 
+                    // Make sure the span actually intersects 'widenedSpan'.  If it 
+                    // does not, just put in an empty length span.  It will get ignored later
+                    // when we walk through this list.
+                    var intersection = span.TextSpan.Intersection(widenedSpan);
+
+                    var newSpan = new ClassifiedSpan(span.ClassificationType,
+                        intersection ?? new TextSpan());
+                    spans[i] = newSpan;
+                }
+            }
+
+            private static void FillInClassifiedSpanGaps(
+                SourceText sourceText, int startPosition,
+                List<ClassifiedSpan> classifiedSpans, ArrayBuilder<ClassifiedSpan> result)
+            {
                 foreach (var span in classifiedSpans)
                 {
+                    // Ignore empty spans.  We can get those when the classification service
+                    // returns spans outside of the range of the span we asked to classify.
+                    if (span.TextSpan.Length == 0)
```

Comment:
Not entirely sure I like the pattern of leaving strange spans in the list for another function to filter out. Any reason not to just widenedSpan to this function and filter it that way?